File: bottom-navigation/src/main/java/it/sephiroth/android/library/bottomnavigation/BadgeProvider.java
package it.sephiroth.android.library.bottomnavigation;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.IdRes;

import java.util.HashSet;

import it.sephiroth.android.library.bottonnavigation.R;
import proguard.annotation.Keep;
import proguard.annotation.KeepClassMembers;

/**
 * Created by alessandro crugnola on 4/12/16.
 * BadgeProvider
 */
@Keep
@KeepClassMembers
public class BadgeProvider {
    private final BottomNavigation navigation;
    private final HashSet<Integer> map = new HashSet<>();
    private final int badgeSize;

    public BadgeProvider(final BottomNavigation navigation) {
        this.navigation = navigation;
        this.badgeSize = navigation.getContext().getResources().getDimensionPixelSize(R.dimen.bbn_badge_size);
    }

    protected Bundle save() {
        Bundle bundle = new Bundle();
        bundle.putSerializable("map", map);
        return bundle;
    }

    @SuppressWarnings ("unchecked")
    public void restore(final Bundle bundle) {
        HashSet<Integer> set = (HashSet<Integer>) bundle.getSerializable("map");
        if (null != set) {
            map.addAll(set);
        }
    }

    /**
     * Returns if the menu item will require a badge
     *
     * @param itemId the menu item id
     * @return true if the menu item has to draw a badge
     */
    public boolean hasBadge(@IdRes final int itemId) {
        return map.contains(itemId);
    }

    Drawable getBadge(@IdRes final int itemId) {
        if (map.contains(itemId)) {
            return newDrawable(itemId, navigation.menu.getBadgeColor());
        }
        return null;
    }

    @SuppressWarnings ("unused")
    protected Drawable newDrawable(@IdRes final int itemId, final int preferredColor) {
        return new BadgeDrawable(preferredColor, badgeSize);
    }

    /**
     * Request to display a new badge over the passed menu item id
     *
     * @param itemId the menu item id
     */
    public void show(@IdRes final int itemId) {
        map.add(itemId);
        navigation.invalidateBadge(itemId);
    }

    /**
     * Remove the currently displayed badge
     *
     * @param itemId the menu item id
     */
    public void remove(@IdRes final int itemId) {
        if (map.remove(itemId)) {
            navigation.invalidateBadge(itemId);
        }
    }
}

File: bottom-navigation/src/main/java/it/sephiroth/android/library/bottomnavigation/BadgeProvider.kt
package it.sephiroth.android.library.bottomnavigation

import android.graphics.drawable.Drawable
import android.os.Bundle
import androidx.annotation.IdRes
import it.sephiroth.android.library.bottonnavigation.R
import java.util.*

/**
 * Created by alessandro crugnola on 4/12/16.
 * BadgeProvider
 *
 * The MIT License
 */

@Suppress("unused")
open class BadgeProvider(private val navigation: BottomNavigation) {
    private val map = HashSet<Int>()
    private val badgeSize: Int = navigation.context.resources.getDimensionPixelSize(R.dimen.bbn_badge_size)

    fun save(): Bundle {
        val bundle = Bundle()
        bundle.putSerializable("map", map)
        return bundle
    }

    @Suppress("UNCHECKED_CAST")
    fun restore(bundle: Bundle) {
        val set = bundle.getSerializable("map")
        if (null != set && set is HashSet<*>) {
            map.addAll(set as HashSet<Int>)
        }
    }

    /**
     * Returns if the menu item will require a badge
     *
     * @param itemId the menu item id
     * @return true if the menu item has to draw a badge
     */
    fun hasBadge(@IdRes itemId: Int): Boolean {
        return map.contains(itemId)
    }

    internal fun getBadge(@IdRes itemId: Int): Drawable? {
        return if (map.contains(itemId)) {
            newDrawable(itemId, navigation.menu!!.badgeColor)
        } else null
    }

    protected open fun newDrawable(@IdRes itemId: Int, preferredColor: Int): Drawable {
        return BadgeDrawable(preferredColor, badgeSize)
    }

    /**
     * Request to display a new badge over the passed menu item id
     *
     * @param itemId the menu item id
     */
    fun show(@IdRes itemId: Int) {
        map.add(itemId)
        navigation.invalidateBadge(itemId)
    }

    /**
     * Remove the currently displayed badge
     *
     * @param itemId the menu item id
     */
    open fun remove(@IdRes itemId: Int) {
        if (map.remove(itemId)) {
            navigation.invalidateBadge(itemId)
        }
    }
}
